src/lib/shim/shim_tls.c

```diff
@@ -0,0 +1,63 @@
+#include "lib/shim/shim_tls.h"
+
+#include "lib/shim/shim.h"
+
+#include <assert.h>
+#include <stdalign.h>
+
+// Needs to be big enough for stack in _vshadow_raw_syscall.
+#define BYTES_PER_THREAD (4096*10 + 1024)
+#define MAX_THREADS 100
+
+// Stores the TLS for a single thread.
+typedef struct ShimThreadLocalStorage {
+    alignas(16) char _bytes[BYTES_PER_THREAD];
+} ShimThreadLocalStorage;
+
+// All TLSs. We could probably make this more dynamic if we need to.
+static ShimThreadLocalStorage _tlss[MAX_THREADS] = {0};
+
+int shimtls_getCurrentIdx();
+int shimtls_takeNextIdx();
+
+// Each ShimTlsVar is assigned an offset in the ShimThreadLocalStorage's.
+// This is the next free offset.
+static size_t _nextByteOffset = 0;
+
+// Index into _tlss.
+static int _tlsIdx = 0;
+
+// Take an unused TLS index, which can be used for a new thread.
+int shimtls_takeNextIdx() {
+    static int next = 0;
+    assert(next < MAX_THREADS);
+    return next++;
+}
+
+// Use when switching threads.
+int shimtls_getCurrentIdx() {
+    static __thread bool idxInitd = false;
+    static __thread int idx;
```

Comment:
Why are C thread-locals allowed here, but not elsewhere in the shim?

Reply:
I have some ideas for how to do this without using the native thread local storage, but it adds some cost and complexity. Doesn't make sense to implement it now, but if we ever do again need to avoid using the native thread-local-storage this is the one place we'll have to change.

Added some comments.